diff --git a/src/http/modules/ngx_http_limit_req_module.c b/src/http/modules/ngx_http_limit_req_module.c
index 6bd3e6a..a41d849 100644
--- a/src/http/modules/ngx_http_limit_req_module.c
+++ b/src/http/modules/ngx_http_limit_req_module.c
@@ -200,6 +200,9 @@ ngx_http_limit_req_handler(ngx_http_request_t *r)
     ngx_http_limit_req_ctx_t    *ctx;
     ngx_http_limit_req_conf_t   *lrcf;
     ngx_http_limit_req_limit_t  *limit, *limits;
+    ngx_table_elt_t             *h;
+    ngx_uint_t                  rate;
+    u_char                      *ra;
 
     if (r->main->limit_req_status) {
         return NGX_DECLINED;
@@ -209,6 +212,7 @@ ngx_http_limit_req_handler(ngx_http_request_t *r)
     limits = lrcf->limits.elts;
 
     excess = 0;
+    rate = 0;
 
     rc = NGX_DECLINED;
 
@@ -252,6 +256,7 @@ ngx_http_limit_req_handler(ngx_http_request_t *r)
                        n, rc, excess / 1000, excess % 1000);
 
         if (rc != NGX_AGAIN) {
+            rate = ctx->rate;
             break;
         }
     }
@@ -268,6 +273,16 @@ ngx_http_limit_req_handler(ngx_http_request_t *r)
                         lrcf->dry_run ? ", dry run" : "",
                         excess / 1000, excess % 1000,
                         &limit->shm_zone->shm.name);
+
+            if ( rate > 0 && (ra = ngx_palloc(r->pool, sizeof("1000"))) != NULL){
+                h = ngx_list_push(&r->headers_out.headers);
+                if ( h != NULL ){
+                    h->hash = 1;
+                    ngx_str_set(&h->key, "Retry-After");
+                    h->value.len = ngx_sprintf(ra, "%ui", (rate >= 1000) ? 1 : 1000/rate) - ra;
+                    h->value.data = ra;
+                }
+            }
         }
 
         while (n--) {
@@ -804,7 +819,7 @@ ngx_http_limit_req_merge_conf(ngx_conf_t *cf, void *parent, void *child)
                                 NGX_LOG_INFO : conf->limit_log_level + 1;
 
     ngx_conf_merge_uint_value(conf->status_code, prev->status_code,
-                              NGX_HTTP_SERVICE_UNAVAILABLE);
+                              NGX_HTTP_TOO_MANY_REQUESTS);
 
     ngx_conf_merge_value(conf->dry_run, prev->dry_run, 0);
 
